Use case-insensitive sorting in tables

okahilak opened the pull request:
Closes #1527

# Definition of Done / Review checklist

## Reviewability
- [ ] link to issue
- [ ] note if PR is on top of other PR
- [ ] note if related change in rems-deploy repo
- [ ] consider adding screenshots for ease of review

## API
- [ ] API is documented and shows up in Swagger UI
- [ ] API is backwards compatible or completely new
- [ ] Events are backwards compatible

## Documentation
- [ ] update changelog if necessary
- [ ] add or update docstrings for namespaces and functions
- [ ] components are added to guide page
- [ ] documentation _at least_ for config options (i.e. docs folder)
- [ ] ADR for major architectural decisions or experiments

## Different installations
- [ ] new configuration options added to rems-deploy repository
- [ ] instance specific translations (i.e. LBR kielivara)

## Testing
- [ ] complex logic is unit tested
- [ ] valuable features are integration / browser / acceptance tested automatically

## Accessibility
- [ ] all icons have the aria-label attribute
- [ ] all fields have a label
- [ ] errors are linked to fields with aria-describedby
- [ ] contrast is checked
- [ ] conscious decision about where to move focus after an action

## Follow-up
- [ ] new tasks are created for pending or remaining tasks
- [ ] no critical TODOs left to implement

opqdonut commented on src/cljs/rems/table.cljs:
```diff
@@ -152,8 +152,8 @@
    (->> rows
         (sort-by #(get-in % [(:sort-column sorting) :sort-value])
```
[nit] you could put str/lower-case here in the keyfn, instead of below in the comparison

okahilak replied:
Done.

okahilak commented on src/cljs/rems/table.cljs:
```diff
@@ -68,14 +67,18 @@
     (assoc m k (make-default m))))
 
-(defn- apply-row-defaults [row]
+(defn apply-row-defaults [row]
```
I had to change this function to public to test it outside the namespace -- otherwise `lein doo once` gives the warning `WARNING: var: rems.table/apply-row-defaults is not public at line 12 test/cljs/rems/test_table.cljs`.

Are there any better solutions?

opqdonut replied:
The usual trick is to refer to privates like `@#'rems.table/apply-row-defaults`, but I'm not sure if that works in cljs.
Another option is to have the test in this file.
Making the function public is fine.

okahilak commented:
The logic for lower-casing was in a slightly wrong place previously, now it fell inside a function that was already being tested. I moved the tests into their own testing namespace (maybe not necessary but I started doing it before realizing that I don't need another deftest but can use the existing ones) and added an explicit test for string and non-string values.

opqdonut commented on test/cljs/rems/test_table.cljs:
```diff
@@ -0,0 +1,81 @@
+                  :filter-value ""
+                  :td [:td ""]}}
+           (apply-row-defaults {:key 123
+                                :foo {:value "FooBar"
```
But if you have `:sort-value "FooBar"`, it won't get lower-cased. Is that intentional?

okahilak replied:
Nope, I'll fix..

okahilak replied:
Hmm, even though that wasn't intentional, I think it could work that way. Currently, :filter-value works analogously, and it makes sense to use the same logic for both, i.e., if someone wants to filter (or sort) explicitly by a non-lower case string, it can be done, but the default is lower-cased.

I added test cases for that for both :filter-value and :sort-value, so that the behavior is more explicit and kind of documented in the test cases.

So, in short, I'd like to keep :filter-value and :sort-value analogous in this regard, and wouldn't want to change the existing behavior of :filter-value without strong enough reasons.

opqdonut replied:
that sounds reasonable
